src/messages_manager.ts

```diff
@@ -20,38 +21,30 @@ export class MessagesManager implements Disposable {
         disposeAll(this.disposables);
     }
 
-    private handleRedraw({ name, args }: EventBusData<"redraw">): void {
+    private async handleRedraw({ name, args }: EventBusData<"redraw">): Promise<void> {
         switch (name) {
             case "msg_show": {
-                const lines: string[] = [];
-                for (const [type, content, clear] of args) {
-                    if (type === "return_prompt") {
-                        // Ignore return_prompt
-                        //
-                        // A note to future readers: return_prompt is sent much more often with ui_messages. It may
-                        // not do what you expect from what :help ui says, so be careful about using these events.
-                        // See: https://github.com/vscode-neovim/vscode-neovim/issues/2046#issuecomment-2144175058
-                        continue;
-                    }
-
-                    if (clear) {
-                        // Remove all stored lines, we will clear the output console on the way out.
-                        lines.splice(0, lines.length);
-                    }
-
-                    const segments = content.map((c) => c[1]);
-                    if (segments.length > 0) {
-                        segments[0] = segments[0].replace(/^\n+/, "");
-                    }
-
-                    lines.push(...segments);
-                }
-
-                if (lines.length >= 2) {
+                const hasReturnPrompt = args.some(([type]) => type === "return_prompt");
+                const msg = args.reduce((acc, [type, content, clear]) => {
+                    // Ignore return_prompt
+                    //
+                    // A note to future readers: return_prompt is sent much more often with ui_messages. It may
+                    // not do what you expect from what :help ui says, so be careful about using these events.
+                    // See: https://github.com/vscode-neovim/vscode-neovim/issues/2046#issuecomment-2144175058
+                    if (type === "return_prompt") return acc;
+                    if (clear) return "";
+                    return acc + content.map((c) => c[1]).join("");
+                }, "");
+                const outputMsg = hasReturnPrompt ? msg.replace(/\n$/, "") : msg;
+
+                this.writeMessage(outputMsg);
+
+                const lineCount = outputMsg.split("\n").length;
+                const cmdheight = (await this.main.client.getOption("cmdheight")) as number;
```

Comment:
Not totally necessary, but it might be nice to say this has to be a minimum of 1, so that every message doesn't show up if someone sets it to zero. I doubt that is ever desirable

Reply:
I don't understand what you mean. By checking `cmdheight` to determine whether to pop up the output, the behavior is basically the same as in nvim. (In nvim, messages are in a window.)

Also, there's no need to include it in the README. If the behavior is consistent with nvim, then there's no need to mention it.

Reply:
cmdheight can take on a value of zero. I'm suggesting we guard against that case

Reply:
A user sets 'cmdheight' to 0 and then complains why the output always pops up?

Reply:
_shrug_ you can take or leave it, it was just a thought I had